src/gevent/_socket3.py

```diff
@@ -308,10 +308,10 @@ def connect_ex(self, address):
         except timeout:
             return EAGAIN
         except error as ex:
-            if type(ex) is error: # pylint:disable=unidiomatic-typecheck
-                return ex.args[0]
-            else:
+            if type(ex) is gaierror: # pylint:disable=unidiomatic-typecheck
```

Comment:
Perusing cpython's `socketmodule.c`'s implementation of `connect_ex`, it appears that the address lookup is done in a separate step/function from the connect call. If *anything* goes wrong in the lookup step, an exception is (re)raised. If *anything* goes wrong in the connect step, the errno is returned.

Looking at address resolution, it appears that it can raise (at least) an OverflowError, a TypeError and possibly other errors that are address-family specific (there's a bunch of code to read there!). So if those errors are raised, they need to be re-raised by us. Here's an overflow error:
```python
>>> import socket
>>> s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
>>> s.setblocking(False)
>>> s.connect_ex(('www.google.com', 65789))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OverflowError: getsockaddrarg: port must be 0-65535.
```

So all of this is to say that the original code was closer to correct: return an errno value, but raise anything else. (With this patch as written, we'd wind up returning a str!) A corrected version *could* look something like:
```python
except error as ex:
  if isinstance(ex, error) and getattr(ex, 'errno', 0):
     return ex.errno
  raise
```

And of course it would need a corresponding test.